plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/OpenShiftDeploymentCleaner.java

```diff
@@ -0,0 +1,139 @@
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import javax.inject.Singleton;
+
+import org.eclipse.che.plugin.openshift.client.exception.OpenShiftException;
+import org.eclipse.che.plugin.openshift.client.kubernetes.KubernetesResourceUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.PodList;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.extensions.Deployment;
+import io.fabric8.kubernetes.api.model.extensions.DoneableDeployment;
+import io.fabric8.kubernetes.api.model.extensions.ReplicaSet;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.dsl.FilterWatchListDeletable;
+import io.fabric8.kubernetes.client.dsl.ScalableResource;
+import io.fabric8.openshift.api.model.Route;
+import io.fabric8.openshift.client.DefaultOpenShiftClient;
+import io.fabric8.openshift.client.OpenShiftClient;
+
+@Singleton
+public class OpenShiftDeploymentCleaner {
+    private static final Logger LOG = LoggerFactory.getLogger(OpenShiftDeploymentCleaner.class);
+
+    public void cleanDeploymentResources(final String deploymentName, final String namespace) throws IOException {
+        scaleDownDeployment(deploymentName, namespace);
+        cleanUpWorkspaceResources(deploymentName, namespace);
+        waitUntilWorkspacePodIsDeleted(deploymentName, namespace);
+    }
+
+    private void scaleDownDeployment(String deploymentName, final String namespace) throws OpenShiftException {
+        try (OpenShiftClient openShiftClient = new DefaultOpenShiftClient()) {
+            ScalableResource<Deployment, DoneableDeployment> deployment = openShiftClient.extensions()
+                           .deployments()
+                           .inNamespace(namespace)
+                           .withName(deploymentName);
+
+            if (deployment != null) {
+                deployment.scale(0, true);
+            }
+        }
+    }
+
+    private void cleanUpWorkspaceResources(final String deploymentName, final String namespace) throws IOException {
+        Deployment deployment = KubernetesResourceUtil.getDeploymentByName(deploymentName, namespace);
+        Service service = KubernetesResourceUtil.getServiceBySelector(OpenShiftConnector.OPENSHIFT_DEPLOYMENT_LABEL, deploymentName, namespace);
+        List<Route> routes = KubernetesResourceUtil.getRoutesByLabel(OpenShiftConnector.OPENSHIFT_DEPLOYMENT_LABEL, deploymentName, namespace);
+        List<ReplicaSet> replicaSets = KubernetesResourceUtil.getReplicaSetByLabel(OpenShiftConnector.OPENSHIFT_DEPLOYMENT_LABEL, deploymentName, namespace);
+
+        try (OpenShiftClient openShiftClient = new DefaultOpenShiftClient()) {
+            if (replicaSets != null && replicaSets.size() > 0) {
```

Comment:
I worry about the order of removal here, since replica sets are controlled by deployments. It shouldn't be a problem since the deployment is scaled to zero first, but I think the deployment should be deleted before the replica set just in case. IIRC there are cases where a deployment will ensure a replica set exists, causing it to maybe be recreated before the deployment is deleted.